patches/server/1052-add-number-format-api.patch

```diff
@@ -0,0 +1,162 @@
++            return NumberFormat.blank();
++        }
++
++        throw new IllegalArgumentException("Unknown format type " + vanilla.getClass());
++    }
++}
+diff --git a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftObjective.java b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftObjective.java
+index 3157f3d2f9ce7af4a763203672817a7f5c7bd4fb..bdd2ed0aa62eea87d356d2229bb7fe41ba293e1d 100644
+--- a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftObjective.java
++++ b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftObjective.java
+@@ -152,6 +152,30 @@ final class CraftObjective extends CraftScoreboardComponent implements Objective
+     }
+     // Paper end
+ 
++    //Paper start - add number format
++    @Override
++    public io.papermc.paper.scoreboard.numbers.NumberFormat getNumberFormat() {
++        net.minecraft.network.chat.numbers.NumberFormat vanilla = this.objective.numberFormat();
++        
++        if(vanilla == null) {
++            return null;
++        }
++
++        return io.papermc.paper.util.PaperScoreboardFormat.asPaper(vanilla);
++    }
++
++
++    @Override
++    public void setNumberFormat(io.papermc.paper.scoreboard.numbers.NumberFormat format) {
++        if(format == null) {
++            this.objective.setNumberFormat(null);
++            return;
++        }
++
++        this.objective.setNumberFormat(io.papermc.paper.util.PaperScoreboardFormat.asVanilla(format));
++    }
++    //Paper end - add number format
++
+     @Override
+     public void unregister() {
+         CraftScoreboard scoreboard = this.checkState();
+diff --git a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftScore.java b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftScore.java
+index be020d6b686c1ad5bd8b7cee0b6050304d434022..e07d8278722123341c9a60e99dccacab37a6c21e 100644
+--- a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftScore.java
++++ b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftScore.java
+@@ -57,6 +57,35 @@ final class CraftScore implements Score {
+         this.objective.checkState().board.getOrCreatePlayerScore(this.entry, this.objective.getHandle()).set(score);
+     }
+ 
++
++    //Paper start - add number format
++    @Override
++    public io.papermc.paper.scoreboard.numbers.NumberFormat getNumberFormat() {
++        net.minecraft.network.chat.numbers.NumberFormat vanilla = this.objective.checkState().board
++            .getPlayerScoreInfo(this.entry, this.objective.getHandle()).numberFormat();
++
++        if(vanilla == null) {
++            return null;
++        }
++
++        return io.papermc.paper.util.PaperScoreboardFormat.asPaper(vanilla);
++    }
++
++
++    @Override
++    public void setNumberFormat(io.papermc.paper.scoreboard.numbers.NumberFormat format) {
++        final net.minecraft.world.scores.ScoreAccess access = this.objective.checkState()
++            .board.getOrCreatePlayerScore(this.entry, this.objective.getHandle());
```

Comment:
Looking into the logic further... This seems not ideal. What do we expect setNumberFormat to do on a score that doesn't exist? Is it right that it creates a score for this entry? I'm not sure it is. One possible solution is to throw an exception on these setters/getters if the score isn't set, can link to Score#isScoreSet to check that first or something. 

Curious for opinions.

Reply:
I think throwing an exception is the best solution. Since that happens when the objective is unregistered as well

Reply:
So my one thing against that is that the /scoreboard command does create an empty score if one doesn't exist if you only change the number format.

Reply:
Sooo we should just leave it like I have it here? How about querying? Do we make that only accessible if there is a score?

Reply:
Ok, so I think the best solution is as follows...
Setters on Score should add the score if it doesn't exist. If we don't do that, we introduce behavior where calling the same methods in different orders behaves differently (setScore, setNumberFormat vs setNumberFormat, setScore).
Getters on Score should not add the score, but return some default value, null, false, 0, the default values for that type. 

And this behavior can be documented in the javadocs

Reply:
So basically leave it like I already had it? I implemented a fix (pushing shortly) because it would throw a npe if the score had not been registered, but other than that, what you're suggesting is what it's already like?

Reply:
Yes, this is now correct.